.NET/Microsoft.Recognizers.Text.DateTime/Utilities/DateContext.cs

```diff
@@ -0,0 +1,81 @@
+    public class DateContext
+    {
+        public int Year { get; set; } = Constants.InvalidYear;
+
+        public DateTimeParseResult ProcessDateEntityParsingResult(DateTimeParseResult originalResult)
+        {
+            if (!IsEmpty())
+            {
+                originalResult.TimexStr = SetTimexWithContext(originalResult.TimexStr);
+                originalResult.Value = ProcessDateEntityResolution((DateTimeResolutionResult)originalResult.Value);
+            }
+
+            return originalResult;
+        }
+
+        public DateTimeResolutionResult ProcessDateEntityResolution(DateTimeResolutionResult resolutionResult)
+        {
+            if (!IsEmpty())
+            {
+                resolutionResult.Timex = SetTimexWithContext(resolutionResult.Timex);
+                resolutionResult.FutureValue = SetDateWithContext((DateObject)resolutionResult.FutureValue);
+                resolutionResult.PastValue = SetDateWithContext((DateObject)resolutionResult.PastValue);
+            }
+
+            return resolutionResult;
+        }
+
+        public DateTimeResolutionResult ProcessDatePeriodEntityResolution(DateTimeResolutionResult resolutionResult)
+        {
+            if (!IsEmpty())
+            {
+                resolutionResult.Timex = SetTimexWithContext(resolutionResult.Timex);
+                resolutionResult.FutureValue = SetDateRangeWithContext((Tuple<DateObject, DateObject>)resolutionResult.FutureValue);
+                resolutionResult.PastValue = SetDateRangeWithContext((Tuple<DateObject, DateObject>)resolutionResult.PastValue);
+            }
+
+            return resolutionResult;
+        }
+
+        public bool IsEmpty()
+        {
+            return this.Year == Constants.InvalidYear;
+        }
+
+        // This method is to ensure the begin date is less than the end date
+        // As DateContext only support common Year as context, so decrease year part of begin date to ensure the begin date is less than end date
+        public DateObject SwiftDateObject(DateObject beginDate, DateObject endDate)
+        {
+            if (beginDate > endDate)
+            {
+                beginDate = beginDate.AddYears(-1);
+            }
+
+            return beginDate;
+        }
+
+        private DateObject SetDateWithContext(DateObject originalDate)
+        {
+            return new DateObject(Year, originalDate.Month, originalDate.Day);
+        }
+
+        private Tuple<DateObject, DateObject> SetDateRangeWithContext(Tuple<DateObject, DateObject> originalDateRange)
+        {
+            var startDate = SetDateWithContext(originalDateRange.Item1);
+            var endDate = SetDateWithContext(originalDateRange.Item2);
+
+            return new Tuple<DateObject, DateObject>(startDate, endDate);
+        }
+
+        private string SetTimexWithContext(string originalTimex)
+        {
+            return originalTimex.Replace(Constants.TimexFuzzyYear, Year.ToString("D4"));
```

Comment:
All timex manipulation code should go to the timex generation code. Let's try to not leave it scattered around.